docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,347 @@
+
+_Fig 2: Proposed design_
+
+### Element: Gardener Custom Metrics Provider Component 
+A new component, named gardener-custom-metrics is added to seed clusters. It periodically scrapes the metrics endpoints
+of all ShootKapi pods on the seed. It implements the K8s custom metrics API and provides K8s metrics specific to
+Gardener, based on custom calculations. The proposed design can naturally be extended to multiple sources of
+input data and a non-volatile cache for acquired data. However, in the scope of this proposal, the only data source
+is the aforementioned metrics scrape, and the calculated values only need to be briefly cached in memory.
+
+A single replica is deployed, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
+
+A kube-apiserver's metrics response measures megabytes in size. To reduce the amount of network traffic, this proposal
+utilizes compressed HTTP responses when scraping metrics. Outside the scope of this proposal, a future enhancement is
+possible, in which a metrics filter component is added as a sidecar to each ShootKapi pod. Each such sidecar would
+mirror the metrics endpoint of its respective ShootKapi, with the only addition of support for filtering based on an
+HTTP request parameter (e.g. `GET /metrics?name=apiserver_request_total` would forward a `GET /metrics` request to the
+ShootKapi, and then would reduce the response to only `apiserver_request_total` counters, before passing it on to the
+caller). 
+
+### Element: New Custom Pod Metric for ShootKapis
+A new `shoot:apiserver_request_total:rate` pod custom metric is made available for each ShootKapi pod on the
+seed. It is provided by the gardener-custom-metrics component. It is the rate of API requests per second, broken down
+by ShootKapi pod.
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
```

Comment:
This here interests me the most (more than the 130 lines essay on what's bad in HVPA with charts and stuff). Do we have numbers/justified reason to think/hope, the requests-based approach will be working? That's the heart of the matter here (core idea) and this paragraph doesn't really say more than it's better than HVPA. What is the basis for this statement (pro the new solution and not contra the old one)?

Reply:
>not discuss HA here and stick to how we do it for all other components

Done. I happened to do it simultaneously to you writing the review, so it ended up just above the review in this thread.

I know that the HA discussion was taxing on the reviewers, but as an author I would not be doing my job if I did not understand all factors influencing the decision. Annoying as that may be, I'll later write a couple of lines summarising my understanding of these factors, so @oliver-goetz  @timebertt get a chance to object not just to the document content, but to my reasoning behind it. I try to avoid design by majority.
<br/>
>I cannot even follow the arguments for non-HA

One clarification: not really arguments, but rather a bunch of factors I wanted taken into consideration. It is not possible for me to truly propose an HA solution, because I do not have access to an exact HA criteria which needs to be met by that solution.

To your question, the factors boil down to: Do we want to pay for the CPU and memory of extra passive replica(s), when that still does not make it possible to check the HA checkbox, because of single-replica VPA? (question is now answered)
<br/>
>When a zone goes down (...) we might need auto-scaling

That's mostly v-scaling though. The average requests/second metric does not change until the affected Kapi is removed from the calculation, so no h-scaling.
It's a problem if VPA's only replica happens to be in the lost AZ.
Fortunately, at some point the HTTP 429s and client retries would likely increase the total request count to a level which triggers h-scaling.  
Unfortunately, due to the difference in resource requirements between gardener-custom-metrics (~200MB) and ShootKapi, occasions on which node shortage would prevent scheduling gardener-custom-metrics, but it's still possible to schedule a new ShootKapi replica, will be rare.
<br/>
>pod eviction time was reduced for our components to 60s

I was thinking about a more aggressive configuration on the gardener-metrics-provider deployment. 30-45 seconds of failed liveness probes + 5 seconds eviction grace. Nonetheless, your main point stands: node provisioning delay is unacceptable wrt HA. That's the one factor necessitating multiple replicas, of which I've been aware from the start. No objections to that one, I just needed to know if there are other factors, so I would be designing from an informed position.

>it's also a checkbox feature

That was provoking my questioning of the extra replica. I don't know how to calculate if we want to pay for extra replicas if that brings us closer to HA, but not all the way to checking the checkbox. From what you say, I understand that if we can't get HA, we want to get as close to HA as possible.

Reply:
>This here interests me the most (...) Do we have numbers

We don't have a formal measure. The test which most closely resembles the proposed solution, was one which manually replicated the arrangement - the driving metric was manually created in Prometeus, and VPA and HPA resources were created for the ShootKapi deployment. Then the cluster was loaded via ClusterLoader2. There were also some similar tests in which a seed was left at its natural load, without purposefully loading it.

I didn't produce numbers from those tests, as I didn't have a good idea for an autoscaling quality metric that's practical enough. Evaluation was practically based on human observation. There are some quantities which are obvious candidates for measuring (e.g. latency, eviction frequency). But to get results which are representative of a real-life workload, a real-life target would need to be used. There are major uncontrolled variables in such test, and to achieve measurement accuracy which exceeds that of plain observation, a relatively large data set would be required.

Measuring comparatively against HVPA would spot differences easily, but those are the differences we already know (e.g. HVPA predictably rolling pods at certain point in its behaviour).

Much of the confidence comes from the natural separation of HPA and VPA's responsibilities and simplicity of the proposed design. In practice, the tests were mostly aimed at looking for surprises, and the results were just... well, unremarkable. **Tests only demonstrated that the proposed approach is viable. The preference for it is based strictly on comparison with HVPA and the other potential alternatives considered during design.**

This topic is revisited in my response to your question about validating effects in the stepwise deployment process.

Reply:
The GEP now specifies a list of quantities to be used when evaluating the fitness of the new autoscaling solution.